ast/cinaps/gen_viewer.ml

```diff
@@ -5,6 +5,25 @@ let string_of_ty ty = Grammar.string_of_ty ~internal:false ty
 let variant_viewer_name cname =
   Ml.id (cname ^ "'const")
 
+let shortcut_viewer_name ~shortcut cname =
+  let _, base_name = String.lsplit2_exn ~on:'_' cname in
+  let prefix =
+    match shortcut.Shortcut.outer_record with
+    | "expression" -> "e"
+    | "pattern" -> "p"
+    | "core_type" -> "t"
+    | "structure_item" -> "str"
+    | "signature_item" -> "sig"
+    | "module_expr" -> "me"
+    | "module_type" -> "mt"
+    | "class_field" -> "cf"
+    | "class_expr" -> "ce"
+    | "class_type" -> "ct"
+    | "class_type_field" -> "ctf"
+    | s -> failwith "No prefix for shortcut: " ^ s
```

Comment:
Could this line cause a failure when trunk adds a new outer record type?

Reply:
Yes it could! It's not the only such line, for instance we have a few `assert false` lying around in places where it's unclear how we should behave, e.g. if we have shortcuts to polymorphic types.

Reply:
I agree in this case we could probably have a default fallback for the shortcut prefix and fix it to something shorter and more meaningful when we actually commit the new versioned viewers.

How about falling back to `"Sc" ^ s`?

Reply:
Here is my thinking: such functions are only necessary for when people want to start migrating their ppx to a newer version of the AST. So we could use a model with two parameters:

1. the highest version we can convert to and from, which we be the max between the higest version in the history of the compiler and the local copy in ppx
2. the highest developers of ppx know about. This would one be bumped manually and could match the version of the ast used in the ppx library itself

we could generate such modules only up to (2). We might then get errors when bumping 2, but that's fine as it would be bumped manually. WDYT?

/cc @ceastlund  as well

Reply:
Yeah I think that's better and more reasonable. We won't be able to generate code for any arbitrary change in the AST anyway.

To sum up, we would generate `conversions` and eventually `versions` and `traverse` as they are relatively simple for up to (1) and `builder` and `viewer` for up to (2). Is that correct?

Reply:
I would only generate `conversions` up to 1, that's the only thing we need to ensure compatibility with trunk. This way, you'd get full support for developing against a new ast version at once, rather than minimal support followed by full support. Indeed, minimal support would quickly be very limited to do anything serious.

Reply:
Perfect, we may leave the `failwith` then!

Reply:
In any case, we can sort that out in a separate PR.

Reply:
I think they two meaningful options for things like this is either fail on unknown, which means certain AST changes will break ppx for that version and force it to update -- which it needs to do anyway -- or to ignore unknown cases and not generate helpers for them, which means new versions work instantly but may need manual help to make them complete. I suspect in the long run neither approach is perfect and it doesn't make a huge difference.

Reply:
I think I prefer the failure mode because it won't let things slip through and we'll always have to deal with them.

Reply:
Having the new versions work instantly but may need manual help has one problem though: in some cases the manual help might mean breaking a bit the API. So if people start using the new functions before we do the manual work, their code might break when we do the manual work.

Reply:
How could that happen? We would only release once the manual work is done and the automated changes (`Conversion`) aren't exposed anyway.

Reply:
I was referring to "to ignore unknown cases and not generate helpers for them, which means new versions work instantly but may need manual help to make them complete" which I interpreted as:

for all versions up to (1), we generate all the modules, using a fail-proof policy that simply skip things we don't know how to name. Then, if things are incomplete or messed up we do a new release of ppx that fixes the generation for the most recent versions of the AST.

So basically it's a best effort policy. It has the advantage that when no manual adjustments are needed, you don't even need to press a button to get the helpers and co for a new version of the AST. But on the other hand, it means that sometimes there will be versions out there of the "ppx" package that will have incomplete APIs, or worse APIs that we don't want at all.

Reply:
The only things that can "fail" should be assigning shorthand. And we should always be able to provide that without breaking compatibility by choosing non-overlapping shorthand for any new types.

Reply:
Agreed. I'm just thinking that a new type might make us reconsider some of the existing shorthands. The current shorthands work well for the current set of types, but with a different set of types we might want to use different shorthands altogether. Maybe it's a bit far-fetched, but if we bump (2) manually at least we always have this choice.